pkg/pod/entrypoint.go

```diff
@@ -0,0 +1,174 @@
+)
+
+const (
+	toolsVolumeName  = "tools"
+	mountPoint       = "/builder/tools"
+	entrypointBinary = mountPoint + "/entrypoint"
+
+	downwardVolumeName     = "downward"
+	downwardMountPoint     = "/builder/downward"
+	downwardMountReadyFile = "ready"
+	ReadyAnnotation        = "tekton.dev/ready"
+	ReadyAnnotationValue   = "READY"
+
+	StepPrefix    = "step-"
+	SidecarPrefix = "sidecar-"
+)
+
+var (
+	// TODO: Generate volumeMount names, to avoid collisions.
+	// TODO: Unexport these vars when Pod conversion is entirely within
+	// this package.
+	ToolsMount = corev1.VolumeMount{
+		Name:      toolsVolumeName,
+		MountPath: mountPoint,
+	}
+	ToolsVolume = corev1.Volume{
+		Name:         toolsVolumeName,
+		VolumeSource: corev1.VolumeSource{EmptyDir: &corev1.EmptyDirVolumeSource{}},
+	}
+
+	// TODO(jasonhall): Signal sidecar readiness by injecting entrypoint,
+	// remove dependency on Downward API.
+	DownwardVolume = corev1.Volume{
+		Name: downwardVolumeName,
+		VolumeSource: corev1.VolumeSource{
+			DownwardAPI: &corev1.DownwardAPIVolumeSource{
+				Items: []corev1.DownwardAPIVolumeFile{{
+					Path: downwardMountReadyFile,
+					FieldRef: &corev1.ObjectFieldSelector{
+						FieldPath: fmt.Sprintf("metadata.annotations['%s']", ReadyAnnotation),
+					},
+				}},
+			},
+		},
+	}
+	DownwardMount = corev1.VolumeMount{
+		Name:      downwardVolumeName,
+		MountPath: downwardMountPoint,
+	}
+)
+
+// OrderContainers returns the specified steps, modified so that they are
+// executed in order by overriding the entrypoint binary.
+//
+// Containers must have Command specified; if the user didn't specify a
+// command, we must have fetched the image's ENTRYPOINT before calling this
+// method, using entrypoint_lookup.go.
+//
+// TODO(jasonhall): Also use entrypoint injection to order sidecar start/stop.
+func OrderContainers(entrypointImage string, orig []corev1.Container) (corev1.Container, []corev1.Container) {
+	toolsInit := corev1.Container{
+		Name:         "place-tools",
+		Image:        entrypointImage,
+		Command:      []string{"cp", "/ko-app/entrypoint", entrypointBinary},
+		VolumeMounts: []corev1.VolumeMount{ToolsMount},
+	}
+
+	steps := orig[:] // defensive copy
+	for i, s := range steps {
+		var argsForEntrypoint []string
+		switch i {
+		case 0:
+			argsForEntrypoint = []string{
+				// First step waits for the Downward volume file.
+				"-wait_file", filepath.Join(downwardMountPoint, downwardMountReadyFile),
+				"-wait_file_content", // Wait for file contents, not just an empty file.
+				// Start next step.
+				"-post_file", filepath.Join(mountPoint, fmt.Sprintf("%d", i)),
+			}
+		default:
+			// All other steps wait for previous file, write next file.
+			argsForEntrypoint = []string{
+				"-wait_file", filepath.Join(mountPoint, fmt.Sprintf("%d", i-1)),
+				"-post_file", filepath.Join(mountPoint, fmt.Sprintf("%d", i)),
+			}
+		}
+
+		cmd, args := s.Command, s.Args
+		if len(cmd) == 0 {
+			panic("oh no bad stuff") // TODO remove
```

Comment:
is this WIP or to be resolved in a follow up PR?

Reply:
Fixed!